### ISSUE TITLE
[Help Wanted] Use viper as the configuration solution and support hot reload for config files

### ISSUE THREAD

**Issue by Starnop** [dragonflyoss/Dragonfly#959]
## Why you need it?

Now, we use different ways to handle the config files for `dfdaemon`, `dfget` and `supernode`. 

Let's take supernode as an example. In order to achieve that  command lines have a higher priority than the configuration file. We should do the three things as follows:
1. load config from config file
2. get cli config which specified by user explicitly
3. integrate the above two configurations which means use cli configurations overrides the configurations from config file

And everything will be simple with [viper](https://github.com/spf13/viper).

In addition, support for reloading a configuration with zero downtime is very useful especially after the project goes online. And [viper](https://github.com/spf13/viper#watching-and-re-reading-config-files) also supports the ability to have your application live read a config file while running. 

So I think we need to do following things:
- [ ] use viper to manage the configurations for `supernode`
- [ ] support hot reload config files for both `dfdaemon` and `supernode`

**Comment by Starnop** [comment]
cc/ @lowzj  @inoc603 @yeya24 

Please take a review. THX. 😄

**Comment by inoc603** [comment]
Agreed. `dfdaemon` is already using viper, and it should be rather easy to support config hot reloading.

**Comment by lowzj** [comment]
agree

**Comment by Starnop** [comment]
@SataQiu Would you like to fix this issue? 😄

**Comment by SataQiu** [comment]
Yes, I‘d like to pick it up.  @Starnop 
But these two days are a bit busy, I will try to fix this issue in the next two days.

**Comment by Starnop** [comment]
@SataQiu I have updated this issue. And would you like to continue to fix it? 😄

**Comment by SataQiu** [comment]
Recently, I was trying to implement hot reload for `dfdaemon`.
But I'm confused about how to deal with `Config` object.
For now, we create the final `Config` object by integrating configurations from different sources(flag and file), then pass this `Config` object to the programs which need it.
It's easy to refresh `Config` object with viper, but how do we tell other components to re-read the configuration. 
I think we might need a global `Config` singleton. Instead of passing the `Config` object everywhere, we should let the program get the global `Config` object through a static method(maybe `config.GetConfig()`). We can then implement hot reload by updating the global `Config` singleton.
Also, I don't think all configuration items support hot reloading. For example, listening ports are difficult to change without restarting the program. Perhaps we should plan which configuration items should support hot reload.
Looking forward to your Suggestions! @yeya24 @inoc603 @Starnop

**Comment by yeya24** [comment]
> /cc @allencloud @lowzj

I agree with that. A global config var may help. I think we should divide the current configuration into two groups first, one group cannot be hot reload and these vars cannot change during runtime like port, log level, I think these vars should be passed via flags. Another group of vars support hot reload, like registry conf, they can only be passed via config files. 

This pattern is very similar to Prometheus, which I think is a good way to deal with configurations.
